### ISSUE TITLE
Error, Missing file, file '/home/reid/autonomous-bittle/vision-based-command/acorns_and_decahedrons_v1/best.pt' 

### ISSUE THREAD

**Issue by Getafix17** [gravesreid/bittle_ros2#2]
Hi Mr. Reid,
You may remember our conversation earlier on the Bittle forum. 
I ran into some errors, if I launch: 
source ~/ros2_ws/install/setup.bash
export ROS_DOMAIN_ID=1
**ros2 launch bittle_ros2 server_joystick_launch.py**
    or
**ros2 launch bittle_ros2 bittle_teleop_robot_launch.py**
The error message indicates that the file '/home/reid/autonomous-bittle/vision-based-command/acorns_and_decahedrons_v1/best.pt' cannot be found. 
[object_detection_subscriber-2]   File "/home/alexf/.local/lib/python3.10/site-packages/torch/serialization.py", line 426, in __init__
[object_detection_subscriber-2]     super().__init__(open(name, mode))
[object_detection_subscriber-2] FileNotFoundError: [Errno 2] No such file or directory: '/home/reid/autonomous-bittle/vision-based-command/acorns_and_decahedrons_v1/best.pt'
[ERROR] [object_detection_subscriber-2]: process has died [pid 4234, exit code 1, cmd '/home/alexf/ros2_ws/install/bittle_ros2/lib/bittle_ros2/object_detection_subscriber --ros-args'].
^C[WARNING] [launch]: user interrupted with ctrl-c (SIGINT)
[rqt_robot_steering-1] [INFO] [1709033807.251245685] [rclcpp]: signal_handler(signum=2)
[INFO] [rqt_robot_steering-1]: process has finished cleanly [pid 4232]

**On the Pi zero I have an issue with the camera** 
**ros2 run usb_cam usb_cam_node_exe --ros-args --params-file src/usb_cam/config/params.yaml**
[ERROR] [1709035097.429339574] [rcl]: Failed to parse global arguments
terminate called after throwing an instance of 'rclcpp::exceptions::RCLInvalidROSArgsError'
  what():  failed to initialize rcl: Couldn't parse params file: '--params-file src/usb_cam/config/params.yaml'. Error: Error opening YAML file, at ./src/parser.c:270, at ./src/rcl/arguments.c:406
[ros2run]: Aborted

**ros2 run v4l2_camera v4l2_camera_node**
[INFO] [1709035227.972630694] [v4l2_camera]: Driver: uvcvideo
[INFO] [1709035227.974873220] [v4l2_camera]: Version: 393493
[INFO] [1709035227.975075343] [v4l2_camera]: Device: USB2.0 PC CAMERA: USB2.0 PC CAM
[INFO] [1709035227.975178957] [v4l2_camera]: Location: usb-3f980000.usb-1
[INFO] [1709035227.975271475] [v4l2_camera]: Capabilities:
[INFO] [1709035227.975362014] [v4l2_camera]:   Read/write: NO
[INFO] [1709035227.975449740] [v4l2_camera]:   Streaming: YES
[INFO] [1709035227.975617429] [v4l2_camera]: Current pixel format: YUYV @ 640x480
[INFO] [1709035227.977124809] [v4l2_camera]: Available pixel formats: 
[INFO] [1709035227.977390018] [v4l2_camera]:   YUYV - YUYV 4:2:2
[INFO] [1709035227.977517543] [v4l2_camera]: Available controls: 
[INFO] [1709035227.977842972] [v4l2_camera]:   Brightness (1) = 128
[INFO] [1709035227.978155169] [v4l2_camera]:   Contrast (1) = 148
[INFO] [1709035227.978454446] [v4l2_camera]:   Saturation (1) = 90
[INFO] [1709035227.978782584] [v4l2_camera]:   Hue (1) = 0
[INFO] [1709035227.979513353] [v4l2_camera]:   Gamma (1) = 4
[INFO] [1709035227.979894782] [v4l2_camera]:   Power Line Frequency (3) = 1
[INFO] [1709035227.980192549] [v4l2_camera]:   Sharpness (1) = 3
[INFO] [1709035227.980677384] [v4l2_camera]:   Backlight Compensation (1) = 1
[INFO] [1709035227.995452692] [v4l2_camera]: Starting camera
[WARN] [1709035228.523307928] [v4l2_camera]: Image encoding not the same as requested output, performing possibly slow conversion: yuv422_yuy2 => rgb8
[INFO] [1709035228.592371249] [v4l2_camera]: using default calibration URL
[INFO] [1709035228.592741425] [v4l2_camera]: camera calibration URL: file:///root/.ros/camera_info/usb2.0_pc_camera:_usb2.0_pc_cam.yaml
[ERROR] [1709035228.607045195] [camera_calibration_parsers]: Unable to open camera calibration file [/root/.ros/camera_info/usb2.0_pc_camera:_usb2.0_pc_cam.yaml]
[WARN] [1709035228.607261383] [v4l2_camera]: Camera calibration file /root/.ros/camera_info/usb2.0_pc_camera:_usb2.0_pc_cam.yaml not found

Kindly advise, as I can seem to figure out if I am missing something. Thank you.

**Comment by gravesreid** [comment]
Hello,

For this error:
FileNotFoundError: [Errno 2] No such file or directory:
'/home/reid/autonomous-bittle/vision-based-command/acorns_
and_decahedrons_v1/best.pt'


You are missing the yolo file best.pt. If you want to use the standard yolo
model, download the .pt ffile from them and update the file path for it. If
you don't care about object detection, update the launch file to open the
video subscriber node.

For the v4l2 node, that error won't prevent video streaming, it is just
complaining that it doesn't have a camera calibration or parameter file.

Best,
Reid

**Comment by Getafix17** [comment]
Hello again Mr Reid, Thanks for the reply. I edited the launch files. The camera is working fine now (without object detection), but the robot does not move using the joystick or the rqt_robot_steering GUI.

**Comment by gravesreid** [comment]
Which launch files are you using?
Which controller are you using?

**Comment by Getafix17** [comment]
I am using a GameSir T1 controller. It is detected by the Ubuntu PC as an Xbox controller.
**ros2 run joy joy_node (terminal1)**
[INFO] [1709086355.141828646] [joy_node]: Opened joystick: Xbox 360 Controller.  deadzone: 0.050000
**ros2 topic echo /joy (terminal2)**
    sec: 1709086418
    nanosec: 27997964
  frame_id: joy
axes:
- -0.0
- -0.0
- 1.0
- -0.0
- -0.0
- 1.0
- 0.0
- 0.0
When I launch **robot_joystic_launch.py** on the Pi and **server_joystick_launch.py** the robot does not respond.
Launching  **bittle_teleop_server_launch.py** and **bittle_teleop_robot_launch.py** The robot doesnt respond to commands from the **rqt_robot_steering** GUI.

**Comment by Getafix17** [comment]
Here is the output for each terminal.
On Pi:
 root@PiBittle:/ros2_ws# cd
root@PiBittle:# source /ros2_ws/install/setup.bash
root@PiBittle:# export ROS_DOMAIN_ID=1
root@PiBittle:# **ros2 launch bittle_ros2 bittle_teleop_robot_launch.py**
[INFO] [launch]: All log files can be found below /root/.ros/log/2024-02-27-12-43-56-056960-PiBittle-75
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [bittle_driver-1]: process started with pid [76]
[INFO] [v4l2_camera_node-2]: process started with pid [78]
[bittle_driver-1] [INFO] [1709037844.220511778] [main_module]: Received a /cmd_vel message!
[bittle_driver-1] [INFO] [1709037844.224915944] [main_module]: Linear Components: [0.000000, 0.000000, 0.000000]
[bittle_driver-1] [INFO] [1709037844.229596778] [main_module]: Angular Components: [0.000000, 0.000000, -0.010000]
On PC:
source /ros2_ws/install/setup.bash
alexf@UbuntuMate:$ export ROS_DOMAIN_ID=1
alexf@UbuntuMate:**ros2 launch bittle_ros2 bittle_teleop_server_launch.py**
[INFO] [launch]: All log files can be found below /home/alexf/.ros/log/2024-02-28-10-52-36-990695-UbuntuMate-37421
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rqt_robot_steering-1]: process started with pid [37422]
[INFO] [video_subscriber-2]: process started with pid [37424]
[video_subscriber-2] [INFO] [1709097825.789374256] [image_subscriber]: Receiving video frame
[video_subscriber-2] [INFO] [1709097830.176353623] [image_subscriber]: Receiving video frame
[video_subscriber-2] [INFO] [1709097830.221801527] [image_subscriber]: Receiving video frame
I get the video feed from the usb cam with both teleop.py and joystick.py

**Comment by gravesreid** [comment]
This makes me wonder if this is a serial communication issue. I have been
using port ttyAMA0, but this requires turning off Bluetooth on the
raspberry pi as detailed in the autonomous bittle readme. For the ros
melodic package, did you change the port in the bittle_driver.py file? If
you didn't, try changing the port in my scripts to match the one you had
working in ros melodic.

**Comment by Getafix17** [comment]
Dear Mr. Graves, Thank you for your patience. I got the robot working, and it works beautifully. I had forgotten to allow permission  
sudo usermod -a -G dialout,tty alexf
That, and I had to disable Bluetooth as per your instructions.

**Comment by gravesreid** [comment]
Great, I'm glad it is working for you!

**Comment by Getafix17** [comment]
Hi Mr. Graves, Sorry to bother you again,
 I downloaded autonomous-bittle from your repo to my '/home/alexf/autonomous-bittle/vision-based-command/acorns_
and_decahedrons_v1/best.pt' 
I got the object detection to work using your best.pt file. 
I updated the server_joystick_launch.py. with your new launch file, and got an error. 
ros2 launch bittle_ros2 server_joystick_launch.py

[INFO] [launch]: All log files can be found below /home/alexf/.ros/log/2024-03-02-09-44-43-005457-Ubuntu22-2965

[INFO] [launch]: Default logging verbosity is set to INFO

[ERROR] [launch]: Caught exception in launch (see debug for traceback): Caught exception when trying to load  file of format [py]: invalid syntax. Perhaps you forgot a comma? (server_joystick_launch.py, line 6)
I have checked the syntax multiple times using ChatGPT, and there are no errors. I replaced the file with the original server_joystick_launch.py, which was working perfectly, but the error still persists. 
I did a fresh install of the bittle_ros2 on another Ubuntu 22.04 Jammy Jellyfish virtual machine, and I got the same error with server_joystick_launch.py.  Strangely,the joystick now works on the Ubuntu Mate virtual machine without problems, but with Ubuntu Mate 22.04 I cannot get the object detection to work. 

The server_launch_file.py (old launch file) that used to work and now gives an error:

from launch import LaunchDescription

from launch_ros.actions import Node



def generate_launch_description():

    return LaunchDescription([

        Node(

            package='joy',

            executable='joy_node',

            name='joy_node',

            output='screen'

        ),

        Node(

            package='bittle_ros2',

            executable='video_subscriber',

        )

    ])
